### PR TITLE
Add a way to dynamically change whether a node identity should be advertised.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
In some cases, a node may need to stop or restart advertising its identity without actually tearing down the relevant CHIPDeviceController.

### CODE REVIEW

**Review by jtung-apple — APPROVED** [review]

**Review by plauric — APPROVED** [review]

**Review by kiel-apple — APPROVED** [review]

File: src/credentials/FabricTable.cpp [hunk 1]
```diff
@@ -2164,4 +2164,17 @@ CHIP_ERROR FabricTable::SetFabricIndexForNextAddition(FabricIndex fabricIndex)
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR FabricTable::SetShouldAdvertiseIdentity(FabricIndex fabricIndex, AdvertiseIdentity advertiseIdentity)
```
Comment (tcarmelveilleux) [inline src/credentials/FabricTable.cpp:2167]: Is there a reason why we are recording this in the fabric table?

File: src/credentials/FabricTable.cpp [hunk 2]
```diff
@@ -2164,4 +2164,17 @@ CHIP_ERROR FabricTable::SetFabricIndexForNextAddition(FabricIndex fabricIndex)
...
+{
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex), CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+    FabricInfo * fabricInfo  = GetMutableFabricByIndex(fabricIndex);
+    bool fabricIsInitialized = (fabricInfo != nullptr) && fabricInfo->IsInitialized();
+    VerifyOrReturnError(fabricIsInitialized, CHIP_ERROR_INVALID_FABRIC_INDEX);
```
Comment (tcarmelveilleux) [inline src/credentials/FabricTable.cpp:2173]: This ternary-ness should be unit tested

File: src/credentials/FabricTable.h [hunk 3]
```diff
@@ -1027,6 +1029,18 @@ class DLL_EXPORT FabricTable
...
+     * @brief Set the advertising behavior for the fabric identified by `fabricIndex`.
+     *
+     * It is the caller's responsibility to actually restart DNS-SD advertising
+     * as needed after updating this state.
+     *
+     * @param fabricIndex - Fabric Index for which to set the label
+     * @param advertiseIdentity - whether the identity for this fabric should be advertised.
+     * @retval CHIP_ERROR_INVALID_FABRIC_INDEX if fabricIndex does not refer to a fabric in the table
+     */
+    CHIP_ERROR SetShouldAdvertiseIdentity(FabricIndex fabricIndex, AdvertiseIdentity advertiseIdentity);
```
Comment (tcarmelveilleux) [inline src/credentials/FabricTable.h:1042]: Please test the getter in FabricInfo in the unit test after several cases of SetShouldAdveritseIdentity.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/credentials/FabricTable.cpp [hunk 4]
```diff
@@ -2164,4 +2164,17 @@ CHIP_ERROR FabricTable::SetFabricIndexForNextAddition(FabricIndex fabricIndex)
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR FabricTable::SetShouldAdvertiseIdentity(FabricIndex fabricIndex, AdvertiseIdentity advertiseIdentity)
+{
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex), CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+    FabricInfo * fabricInfo  = GetMutableFabricByIndex(fabricIndex);
```
Comment (kiel-apple) [inline src/credentials/FabricTable.cpp:2171]:
```suggestion
    FabricInfo * fabricInfo = GetMutableFabricByIndex(fabricIndex);
```

maybe this was intentional for alignment?

File: src/credentials/FabricTable.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/credentials/FabricTable.cpp:2167]: That's pre-existing, but fundamentally the reason is that DnssdServer::AdvertiseOperational examines the fabric table to decide what it wants to do.

File: src/credentials/FabricTable.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply src/credentials/FabricTable.cpp:2173]: For what it's worth, I am not sure that the `!fabricInfo->IsInitialized()` case is ever reachable in practice.  But other consumers of `GetMutableFabricByIndex` seem to worry about it...

File: src/credentials/FabricTable.h (hunk 3 above)
Comment (bzbarsky-apple) [reply src/credentials/FabricTable.h:1042]: Can do.

File: src/credentials/FabricTable.cpp (hunk 4 above)
Comment (bzbarsky-apple) [reply src/credentials/FabricTable.cpp:2171]: It's enforced by restyled.

**Pull request merged**